### PR TITLE
Monster spawn

### PR DESCRIPTION
**Pull request by Cr3ativeZen**

### CODE REVIEW

File: apps/gameserver/include/Character.h [hunk 1]
```diff
@@ -13,14 +13,15 @@ class Character : public bango::space::quad_entity
 {
 public:
     typedef std::uint32_t id_t;
+    static id_t g_max_id;
```
Comment (lafreak) [inline apps/gameserver/include/Character.h:16]: This should not be public. ID generation should be hidden inside Character class and not exposed to other classes.

Comment (lafreak) [reply apps/gameserver/include/Character.h:16]: To achieve it, I suggest creating new public method: `AssignNewId` which in future will be threadsafe (my task) and will be used inside Character constructor and spawner.

File: apps/gameserver/include/Monster.h [hunk 2]
```diff
@@ -74,11 +75,15 @@ struct InitMonster : public bango::processor::db_object<InitMonster>
 class Monster : public Character
 {
     const std::unique_ptr<InitMonster>& m_init;
+    bango::utils::time::point m_death_time;
 
 public:
     Monster(const std::unique_ptr<InitMonster>& init, int x, int y, int map=0);
     ~Monster();
 
+    void PrepareMonsterToSpawn(int new_x, int new_y, int map);
```
Comment (lafreak) [inline apps/gameserver/include/Monster.h:84]:
```suggestion
    void Respawn(int new_x, int new_y, int map);
```

File: apps/gameserver/include/Spawn.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+struct GenMonster : public bango::processor::db_object<GenMonster>
+{
+    struct RectXY
+    {
+        int X1 = 0, Y1 = 0, X2 = 0, Y2 = 0;
+        int GetRandomX() const;
+        int GetRandomY() const;
+
+    };
+    std::uint32_t MonsterIndex = 0, Map = 0, Index = 0, Amount = 0, SpawnCycle = 0;
```
Comment (lafreak) [inline apps/gameserver/include/Spawn.h:24]:
```suggestion
    std::uint32_t MonsterIndex = 0, Map = 0, Area= 0, Amount = 0, SpawnCycle = 0;
```
Why not call it as it is inside config file?

File: apps/gameserver/include/World.h [hunk 4]
```diff
@@ -129,10 +130,12 @@ class World
     typedef std::unordered_map<Character::id_t, Player*>                    PlayerContainer;
     typedef std::unordered_map<Character::id_t, std::shared_ptr<Monster>>   MonsterContainer;
     typedef std::unordered_map<Character::id_t, std::shared_ptr<NPC>>       NpcContainer;
+    typedef std::unordered_map<Character::id_t, std::shared_ptr<Spawn>>     SpawnContainer;
```
Comment (lafreak) [inline apps/gameserver/include/World.h:133]: Is there any reason why we should keep it inside a map?
I don't think we will ever need to identify spawn by id.
It does not inherit from Character.
Could it be std::list or any container which is not key value?

File: apps/gameserver/include/World.h [hunk 5]
```diff
@@ -179,12 +182,15 @@ class World
             Add(std::make_shared<NPC>(init.second));
     }
 
+    static void SpawnGenMonster();
```
Comment (lafreak) [inline apps/gameserver/include/World.h:185]:
```suggestion
    static void SpawnMonsters();
```

File: apps/gameserver/include/World.h [hunk 6]
```diff
@@ -179,12 +182,15 @@ class World
...
+
     static void Cleanup()
     {
         std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
         Get().m_players.clear();
         Get().m_monsters.clear();
         Get().m_npcs.clear();
+        Get().m_spawns.clear();
```
Comment (lafreak) [inline apps/gameserver/include/World.h:193]: Here we remove monsters before we remove spawns. Is it safe? Will spawner keep respawning monsters because monsters got deleted? What is the expected behavior here?

File: apps/gameserver/src/Monster.cpp [hunk 7]
```diff
@@ -63,24 +63,48 @@ packet Monster::BuildMovePacket(std::int8_t delta_x, std::int8_t delta_y, std::i
...
-            World::Add(std::make_shared<RegularMonster>(InitMonster::DB().at(index), x, y, map));
+            World::Add(std::make_shared<RegularMonster>(init_monster, x, y, map));
             break;
         }
     }
+}
+
+void Monster::SetNewID()
+{
+    m_id = Character::g_max_id++;
```
Comment (lafreak) [inline apps/gameserver/src/Monster.cpp:91]: Why moved declaration out of this function? I think we should minimize access to that variable and only have access to it here.

File: apps/gameserver/src/Monster.cpp [hunk 8]
```diff
@@ -63,24 +63,48 @@ packet Monster::BuildMovePacket(std::int8_t delta_x, std::int8_t delta_y, std::i
...
+}
+
+void Monster::PrepareMonsterToSpawn(int new_x, int new_y, int map)
```
Comment (lafreak) [inline apps/gameserver/src/Monster.cpp:94]:
```suggestion
void Monster::RestoreInitialState(int new_x, int new_y, int map)
```

Comment (lafreak) [reply apps/gameserver/src/Monster.cpp:94]: Why is `int map` here?

File: apps/gameserver/src/RegularMonster.cpp [hunk 9]
```diff
@@ -2,13 +2,15 @@
...
 #include <bango/network/packet.h>
+#include <bango/utils/time.h>
 #include <inix.h>
 
 using namespace bango::network;
 
 void RegularMonster::Die()
 {
     SetGState(CGS_KO);
+    SetDeathTime(bango::utils::time::now());
```
Comment (lafreak) [inline apps/gameserver/src/RegularMonster.cpp:13]: Why do RegularMonsters need to have death time set?

File: apps/gameserver/src/Spawn.cpp [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+void Spawn::Tick()
+{
+    if((time::now() - m_next_spawn_cycle).count() > GetSpawnCycle())
+        return;
+
+    for(auto& monster : m_area_monsters)
+    {
+        auto monster_lock = monster->Lock();
```
Comment (lafreak) [inline apps/gameserver/src/Spawn.cpp:52]: Here you lock because you want to check if monster is ready to be respawned (CGS_KO).
I don't think you really need to lock.

In the context of monster _just_ got killed and you ask too quick for CGS_KO, its not a problem. We will not respawn but the next tick iteration will respawn it, right? We can live with that.

In the context of we checking CGS_KO and we are afraid that when we perform respawn some other thread will change state (remove CGS_KO) - can it really happen? I think the only thread that can remove CGS_KO should be this tick thread during respawn. Other thread should NEVER do that, right?

We need to ensure no other thread edits monster during respawn. How to do that?
Which other threads can edit? Player recv packets. For example player->monster attack.
If we ensure that monster is already CGS_KO during attack packet we SHOULD reject attack. This will ensure us that nothing can write to it beside spawner.

To summarize, CGS_KO can be our safe flag that there is only one thread that can edit monster - Ticker.

File: apps/gameserver/src/Spawn.cpp [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    World::Add(monster);
+}
+
+void Spawn::CreateSpawn()
+{
+    const auto& init_monster = InitMonster::DB().at(GetMonsterIndex());
+    m_area_monsters.resize(GetAmount());
+    for (int i = 0; i < GetAmount(); i++)
+    {
+        switch (init_monster->Race)
```
Comment (lafreak) [inline apps/gameserver/src/Spawn.cpp:72]: This switch duplicates switch inside Monster::Summon. This is not maintainable in the long run - what if we add 5 more monster kinds - do we want to remember to add to support all types in both places. I dont think we need to.

My suggestion is to separate the Monster builder method and spawn method.

Monster::CreateMonster could be the builder and create monster based on init,x,y,map. Create but not add to the world. We could have the switch in there.

Other logic should be separated and re-use Monster::CreateMonster.

Here we could call Monster::CreateMonster, check if it is not nullptr (error during creation) and insert into the world and spawner container.
Other place (on demand Summon - /mob or boss spawn) could also reuse Monster::CreateMonster and add to the world in separate Monster::Summon method.

File: apps/gameserver/src/World.cpp [hunk 12]
```diff
@@ -326,4 +329,21 @@ void WorldMap::WriteOnMap(const packet& p)
 
     for (const auto& pair : Players())
         ((Player*)pair.second)->write(p);
+}
+
+void World::SpawnGenMonster()
+{
+    for(const auto& init : GenMonster::DB())
+    {
+        auto new_spawn = std::make_shared<Spawn>(GenMonster::DB().at(init.second->Index));
```
Comment (lafreak) [inline apps/gameserver/src/World.cpp:338]: Here we iterate over GenMonster entries (`init` variable containing key and value pair (`area` and `GenMonster` ptr)).
For each entry, we retrieve the area (`init.second->Index`) and use it to get ` GenMonster` ptr. Why? We already have the ptr inside `init` variable. `init.second` is the pointer.

File: apps/gameserver/src/World.cpp [hunk 13]
```diff
@@ -326,4 +329,21 @@ void WorldMap::WriteOnMap(const packet& p)
...
+        Get().m_spawns.insert(std::make_pair(init.second->Index, new_spawn));
+    }
+}
+
+void World::ForEachSpawn(const std::function<void(Spawn&)>& callback)
+{
+    std::lock_guard<std::recursive_mutex> lock(Get().m_entities_rmtx);
```
Comment (lafreak) [inline apps/gameserver/src/World.cpp:345]: We currently do not need to lock. The container does not change in runtime.
It will probably change if we add hot reload at runtime (inix supports that?)

File: apps/gameserver/include/Character.h (hunk 1 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/Character.h:16]: done

File: apps/gameserver/include/Spawn.h (hunk 3 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/Spawn.h:24]: done

File: apps/gameserver/src/World.cpp (hunk 12 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/World.cpp:338]: ups, done

File: apps/gameserver/include/World.h (hunk 4 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/World.h:133]: done

File: apps/gameserver/include/World.h (hunk 5 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/World.h:185]: done

File: apps/gameserver/src/Monster.cpp (hunk 7 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Monster.cpp:91]: done

File: apps/gameserver/src/Monster.cpp (hunk 8 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Monster.cpp:94]: done

File: apps/gameserver/src/World.cpp (hunk 13 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/World.cpp:345]: Inix does not support that i think, but it's an useful feature and we should have it later on. I will remove lock_guard for now.

File: apps/gameserver/src/RegularMonster.cpp (hunk 9 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/RegularMonster.cpp:13]: done

File: core/processor/include/bango/processor/db.h [hunk 14]
```diff
@@ -5,6 +5,7 @@
 #include <map>
 #include <functional>
```
[the remark this answers is not in the document]
Comment (Cr3ativeZen) [reply core/processor/include/bango/processor/db.h:6]: ups, done

File: apps/gameserver/src/Spawn.cpp (hunk 10 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp:52]: reload at spawn time might be an issue later on.
i will remove lock for now

File: apps/gameserver/include/World.h (hunk 6 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/World.h:193]: It might be a case, and mob could get added via World::Add back to m_monsters.

File: apps/gameserver/src/Spawn.cpp (hunk 11 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp]: done

File: apps/gameserver/src/Monster.cpp [hunk 15]
```diff
@@ -63,48 +63,30 @@ packet Monster::BuildMovePacket(std::int8_t delta_x, std::int8_t delta_y, std::i
     return packet();
 }
 
-void Monster::SummonMonster(std::uint32_t index, std::int32_t x, std::int32_t y, std::int32_t map)
+void Monster::Summon(std::uint32_t index, std::int32_t x, std::int32_t y, std::int32_t map)
+{
+    World::Add(Monster::CreateMonster(index, x, y, map));
```
Comment (lafreak) [inline apps/gameserver/src/Monster.cpp:68]: Error handling is missing - CreateMonster can return nullptr

File: apps/gameserver/include/Character.h [hunk 16]
```diff
@@ -39,15 +39,15 @@ class Character : public bango::space::quad_entity
...
-        m_id = g_max_id++;
+        AssignNewId();
     }
 
     constexpr static std::uint8_t PLAYER    =0;
     constexpr static std::uint8_t MONSTER   =1;
     constexpr static std::uint8_t NPC       =2;
     constexpr static std::uint8_t LOOT      =3;
 
+    void AssignNewId();
```
Comment (lafreak) [inline apps/gameserver/include/Character.h:50]: Could this be protected? So Monster class can use it but no other classes can.

File: apps/gameserver/include/World.h [hunk 17]
```diff
@@ -179,10 +182,13 @@ class World
             Add(std::make_shared<NPC>(init.second));
     }
 
+    static void SpawnMonster();
```
Comment (lafreak) [inline apps/gameserver/include/World.h:185]:
```suggestion
    static void SpawnMonsters();
```

File: apps/gameserver/src/Spawn.cpp [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+void Spawn::Tick()
+{
+    if((time::now() - m_next_spawn_cycle).count() > GetSpawnCycle())
+        return;
+
+    for(auto& monster : m_area_monsters)
+    {
+        if(monster->IsGState(CGS_KO))
```
Comment (lafreak) [inline apps/gameserver/src/Spawn.cpp:52]: Add note here that we have thread-safety by making sure we always call RemoveDeadMonsters and Spawn::Tick in the same thread and that Spawn::Tick is called **after** RemoveDeadMonsters.

File: apps/gameserver/src/Spawn.cpp [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+            RespawnOnWorld(monster);
+    }
+
+    SetNextSpawnCycle();
+}
+
+void Spawn::RespawnOnWorld(std::shared_ptr<Monster> monster)
+{
+    monster->RestoreInitialState(GetRandomX(), GetRandomY());
+    World::Add(monster);
```
Comment (lafreak) [inline apps/gameserver/src/Spawn.cpp:62]: World add/remove API is missing check if the monster with given ID already exist. It's not an issue here but in the future it will be important to catch bugs early. Add TODO.

File: apps/gameserver/include/Spawn.h [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+class Spawn
+{
+public:
+
+    Spawn(const std::unique_ptr<GenMonster>& init);
+
+    void Tick();
+private:
+    void RespawnOnWorld(std::shared_ptr<Monster> monster);
```
Comment (lafreak) [inline apps/gameserver/include/Spawn.h:39]: This can be const

File: apps/gameserver/src/Spawn.cpp (hunk 18 above)
Comment (lafreak) [inline apps/gameserver/src/Spawn.cpp:52]: This call can crash if we end up in line 72.

```

    for (int i = 0; i < GetAmount(); i++)
    {
        auto monster = Monster::CreateMonster(GetMonsterIndex(), GetRandomX(), GetRandomY(), GetMap());
        if(!monster)
            spdlog::error("Monster was not created correctly.");
        m_area_monsters.at(i) = monster;  // HERE WE POSSIBLY PUT nullptr
        World::Add(monster);
    }
```

and when we put that nullptr and we iterate over vector here in tick we will dereference nullptr.

File: apps/gameserver/src/Spawn.cpp [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+void Spawn::CreateSpawn()
+{
+    const auto& init_monster = InitMonster::DB().at(GetMonsterIndex());
+    m_area_monsters.resize(GetAmount());
+    for (int i = 0; i < GetAmount(); i++)
+    {
+        auto monster = Monster::CreateMonster(GetMonsterIndex(), GetRandomX(), GetRandomY(), GetMap());
```
Comment (lafreak) [inline apps/gameserver/src/Spawn.cpp:71]: As far I see the transaction policy for CreateMonster method is that we always return valid ptr. The only exit from this function is by throwing an exception. We should use that and catch that exception instead of calling `if(!monster)`

Comment (lafreak) [reply apps/gameserver/src/Spawn.cpp:71]: We should not add into the list if the exception happen

Comment (lafreak) [reply apps/gameserver/src/Spawn.cpp:71]: We also should not add nullptr to the world

File: apps/gameserver/include/World.h (hunk 17 above)
Comment (lafreak) [reply apps/gameserver/include/World.h:185]: Even SpawnMonsters is misleading in this case.
This is responsible for two things:
- creating spawns (the containers)
- spawns monsters related to those spawns

My final suggestion is: `CreateSpawns`

Comment (lafreak) [reply apps/gameserver/include/World.h:185]: Or even better: `CreateSpawnsAndSpawnMonsters`

File: apps/gameserver/src/Spawn.cpp [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+
+using namespace bango::utils;
+
+Spawn::Spawn(const std::unique_ptr<GenMonster>& init)
+    :m_init(init)
+{
+    if(InitMonster::DB().find(GetMonsterIndex()) != InitMonster::DB().end())
+        CreateSpawn();
+    else
+        spdlog::error("Monster ID does not exist in InitMonster {}", GetMonsterIndex()); // TODO: object shouldn't be created.
```
Comment (lafreak) [inline apps/gameserver/src/Spawn.cpp:21]: This can easily be done in World::CreateSpawnsAndSpawnMonsters method.

```
void World::CreateSpawnsAndSpawnMonsters()
{
    for(const auto& init : GenMonster::DB())
    {
        if (InitMonster::DB().count(init.second->MonsterIndex) == 0)
        {
             log error;
             continue;
         }
        Get().m_spawns.emplace_back(std::make_shared<Spawn>(init.second));
    }
}
```

Comment (lafreak) [reply apps/gameserver/src/Spawn.cpp:21]: Then we can get rid of the if in line 18.

**Review by lafreak — CHANGES_REQUESTED** [review]
You are 5 commits away from master. Please rebase your changes.

File: apps/gameserver/include/World.h (hunk 17 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/World.h:185]: done

File: apps/gameserver/include/Character.h (hunk 16 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/Character.h:50]: done

File: apps/gameserver/src/Spawn.cpp (hunk 18 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp:52]: done

File: apps/gameserver/src/Spawn.cpp (hunk 19 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp:62]: done

File: apps/gameserver/include/Spawn.h (hunk 20 above)
Comment (Cr3ativeZen) [reply apps/gameserver/include/Spawn.h:39]: done

File: apps/gameserver/src/Spawn.cpp (hunk 22 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp:21]: done

File: apps/gameserver/src/Spawn.cpp (hunk 21 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp:71]: done

Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp:71]: done

File: apps/gameserver/src/Spawn.cpp (hunk 18 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp:52]: try catch added instead, Monster::CreateMonster aka std::make_shared will never return nullptr.

File: apps/gameserver/src/Spawn.cpp (hunk 21 above)
Comment (lafreak) [reply apps/gameserver/src/Spawn.cpp:70]: Add proper formatting (add indent inside try block)

Do not use resize and at - you are still able to leave nullptr in the m_area_monsters and then dereference it in Spawn::Tick which is unsafe

File: apps/gameserver/src/World.cpp [hunk 23]
```diff
@@ -326,4 +330,25 @@ void WorldMap::WriteOnMap(const packet& p)
...
         ((Player*)pair.second)->write(p);
+}
+
+void World::CreateSpawnsAndSpawnMonsters()
+{
+    for(const auto& init : GenMonster::DB())
+    {
+        if (InitMonster::DB().count(init.second->MonsterIndex) == 0)
+        {
+            spdlog::error("Monster ID does not exist in InitMonster {}", init.second->MonsterIndex);
```
Comment (lafreak) [inline apps/gameserver/src/World.cpp:341]: Add spawn id (area) into the log as well for better debugging

Comment (Cr3ativeZen) [reply apps/gameserver/src/World.cpp]: done

File: apps/gameserver/src/Spawn.cpp (hunk 21 above)
Comment (Cr3ativeZen) [reply apps/gameserver/src/Spawn.cpp]: my bad...
done

**Review by lafreak — APPROVED** [review]
